onnxscript/rewriter/generic_pattern.py

```diff
@@ -294,6 +294,78 @@ def make_node(
         return proto
 
 
+class PatternMatchResult:
+    """Stores information about a match if a match was successful."""
+
+    def __init__(
+        self,
+        pattern: "GenericPattern",
+        model_nodes: typing.Sequence[oir.Node],
+        pattern_nodes: typing.Sequence[oir.Node],
+        pattern_input_names: typing.Sequence[str],
+        pattern_output_names: typing.Sequence[str],
+    ):
+        assert len(model_nodes) == len(pattern_nodes), (
+            f"The number of matched nodes in the model {len(model_nodes)} "
+            f"should the same as the number of nodes in the pattern "
+            f"{len(pattern_nodes)}."
+        )
+        self.pattern = pattern
+        self.model_nodes = model_nodes
+        self.pattern_nodes = pattern_nodes
+        self.pattern_input_names = pattern_input_names
+        self.pattern_output_names = pattern_output_names
+        self.kwargs = {}
+
+        matched_pattern_to_model_name: dict[str, str] = {}
+        for gn, pn in zip(model_nodes, pattern_nodes):
+            assert (
+                gn.op_type == pn.op_type
+            ), f"Unexpected type mismatch {gn.op_type!r} != {pn.op_type!r}"
+            assert len(gn.input_names) == len(
+                pn.input_names
+            ), f"Unexpected number of inputs for type {gn.op_type}"
+            for a, b in zip(gn.input_names, pn.input_names):
+                if b == "":
+                    # optional input or not an interesting input
+                    continue
+                if b in matched_pattern_to_model_name:
+                    assert matched_pattern_to_model_name[b] == a, (
+                        f"Ambiguities, pattern name {b!r} means "
+                        f"{a!r} or {matched_pattern_to_model_name[b]}"
+                    )
+                else:
+                    matched_pattern_to_model_name[b] = a
+
+            assert len(gn.output_names) == len(
+                pn.output_names
+            ), f"Unexpected number of outputs for type {gn.op_type}"
+            for a, b in zip(gn.output_names, pn.output_names):
+                if b == "":
+                    # Only final outputs are interesting.
+                    continue
+                assert a != "", f"{a!r} cannot be optional"
+                if b in matched_pattern_to_model_name:
+                    assert matched_pattern_to_model_name[b] == a, (
```

Comment:
So, these asserts mean that the pattern-matcher should have stopped earlier (and skipped the match) in the absence of bugs ... is that right?

Reply:
A repeated variable X (representing a value) does have a valid interpretation: for example, if I want to rewrite "X*A + X*B" into "X*(A+B)". The meaning is that all occurrences of X represent the same value. My question really is that this check will happen earlier (in the pattern-matching algorithm) ... so if we see a subgraph in the model that looks like "W*A + V*B", then the pattern-matcher will fail when the second-occurrence of X is matched against V (which is not same as W). I am saying this is not an error that should cause a runtime-exception, but it is a failed-match that will be skipped.

Reply:
Created issue #1386